src/Components/Authentication/__tests__/Desktop/LoginForm.test.tsx

```diff
@@ -80,6 +82,142 @@ describe("LoginForm", () => {
+        (values, actions) => {
+          actions.setStatus({ error: "some password error" })
+        }
+      )
+
+      const wrapper = getWrapper()
+
+      const inputEmail = wrapper
+        .find(QuickInput)
+        .filterWhere(el => el.prop("name") === "email")
+        .instance() as QuickInput
+      inputEmail.onChange(ChangeEvents.email)
+
+      const inputPass = wrapper
+        .find(QuickInput)
+        .filterWhere(el => el.prop("name") === "password")
+        .instance() as QuickInput
+      inputPass.onChange(ChangeEvents.password)
+
+      const form = wrapper.find(`Formik`)
+      form.simulate("submit")
+
+      await flushPromiseQueue()
+      wrapper.update()
+
+      expect(wrapper.html()).toMatch("some password error")
+
+      expect(props.handleSubmit.mock.calls[0][0]).toEqual({
+        email: "email@email.com",
+        password: "password",
+      })
+    })
+
+    it("shows 2FA one-time password prompt", async () => {
+      ;(props.handleSubmit as jest.Mock).mockImplementation(
+        (values, actions) => {
+          if (!values.otp_attempt) {
+            actions.setStatus({
+              error: "missing two-factor authentication code",
+            })
+          } else if (values.otp_attempt !== "123456") {
+            actions.setStatus({
+              error: "invalid two-factor authentication code",
+            })
+          } else {
+            actions.setStatus(null)
+          }
+        }
+      )
+
+      const wrapper = getWrapper()
+
+      const inputEmail = wrapper
+        .find(QuickInput)
+        .filterWhere(el => el.prop("name") === "email")
+        .instance() as QuickInput
+      inputEmail.onChange(ChangeEvents.email)
+
+      const inputPass = wrapper
+        .find(QuickInput)
+        .filterWhere(el => el.prop("name") === "password")
+        .instance() as QuickInput
+      inputPass.onChange(ChangeEvents.password)
+
+      const form = wrapper.find(`Formik`)
+      form.simulate("submit")
+
+      await flushPromiseQueue()
+      wrapper.update()
+
+      expect(wrapper.html()).not.toMatch(
+        "missing two-factor authentication code"
+      )
+      expect(props.handleSubmit.mock.calls[0][0]).toEqual({
+        email: "email@email.com",
+        password: "password",
+      })
+
+      const inputOtp = wrapper
+        .find(QuickInput)
+        .filterWhere(el => el.prop("name") === "otp_attempt")
+        .instance() as QuickInput
+      inputOtp.onChange(ChangeEvents.invalidOtpAttempt)
+      form.simulate("submit")
+
+      await flushPromiseQueue()
+      wrapper.update()
+
+      expect(wrapper.html()).toMatch("invalid two-factor authentication code")
+      expect(props.handleSubmit.mock.calls[1][0]).toEqual({
+        email: "email@email.com",
+        password: "password",
+        otp_attempt: "111111",
+      })
+
+      inputOtp.onChange(ChangeEvents.otpAttempt)
+      form.simulate("submit")
+
+      await flushPromiseQueue()
```

Comment:
✨ 

Good to know about this util

Not gonna pretend I fully understood, but I found it helpful to read [David's explanation here](https://github.com/artsy/reaction/pull/1694#discussion_r241712877)